### PR TITLE
[WIP] Speaker invitation

### PR DESCRIPTION
**Pull request by rixx**
This PR closes/references issue #155. It does so by adding an `invitation_token` field to submissions, and adding helper views/methods to either share the link directly or via email.

## How Has This Been Tested?
Tests are still tbd.

## Screenshots (if appropriate):
2017-10-17-143210_1283x436_scrot
2017-10-17-143228_1279x845_scrot
2017-10-17-213937_1386x526_scrot

## Checklist
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: src/pretalx/cfp/templates/cfp/event/user_submission_invitation.html [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+        {% trans "Current state of your submission:" %}
+        {% include "cfp/event/fragment_state.html" with state=submission.state %}
+    </p>
+    <div class="alert alert-info">
+        {% blocktrans trimmed %}
+            Invite another speaker to your submission here. Instead, you can also send them
+            this link:
+        {% endblocktrans %}
+        <br>
+        {{ invite_url }}
```
Comment (luto) [inline src/pretalx/cfp/templates/cfp/event/user_submission_invitation.html]: Maybe make this link clickable, so users can do RightClick=>CopyLink?

File: src/pretalx/cfp/templates/cfp/event/user_submission_edit.html [hunk 2]
```diff
@@ -8,6 +8,26 @@
...
     </p>
+    <h5>
+        <form class="add-speaker" action="{{ submission.urls.invite }}">
+            <div class="form-group form-inline">
+                {% blocktrans trimmed count count=submission.speakers.count %}
+                    Speaker:
+                    {% plural %}
+                    Speakers:
+                {% endblocktrans %}
+                {{ submission.display_speaker_names }},
```
Comment (luto) [inline src/pretalx/cfp/templates/cfp/event/user_submission_edit.html]: As a speaker, can I somehow see speakers, who have been invited? Can I access the link again in case they lost it? Can I resend the mail? As a solution, also list all invited speakers here (gray, with tooltip); clicking an invited speaker shows the new-invitation-mail page again.

File: src/pretalx/submission/models/submission.py [hunk 3]
```diff
@@ -12,6 +12,10 @@
 from pretalx.mail.context import template_context_from_submission
 
 
+def generate_invite_code(length=32):
+    return get_random_string(length=length, allowed_chars=Submission.CODE_CHARSET)
```
Comment (luto) [inline src/pretalx/submission/models/submission.py]: Note that the orga team invite also generates tokens for invite links. There are now at least three places in pretalx, which do that. Maybe think about centralizing this in the future.

File: src/pretalx/cfp/templates/cfp/event/user_submission_invitation.html [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    </div>
+    <form action="" method="post">
+        {% csrf_token %}
+        {% bootstrap_form form %}
+        <div class="row">
+            <div class="col-md-3 offset-md-9">
+                <button type="submit" class="btn btn-block btn-success btn-lg">
```
Comment (luto) [inline src/pretalx/cfp/templates/cfp/event/user_submission_invitation.html]: "Send" is great! How about "Cancel"? 😁

File: src/pretalx/cfp/templates/cfp/event/user_submission_edit.html [hunk 5]
```diff
@@ -8,6 +8,26 @@
...
+                <div class="input-group">
+                    <input name="email" class="form-control form-control-sm" placeholder="another.speaker@mail.org">
```
Comment (luto) [inline src/pretalx/cfp/templates/cfp/event/user_submission_edit.html]: The form doesn't complain if the user inputs something else, like a speaker name. The next page shows the data entered here as "Speaker", not as "Speaker E-Mail". Maybe display a yellow warning thingy on the next page.

File: src/pretalx/cfp/templates/cfp/event/user_submission_invitation.html [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+        </a>
+    </h2>
+    <p>
+        {% trans "Current state of your submission:" %}
+        {% include "cfp/event/fragment_state.html" with state=submission.state %}
+    </p>
+    <div class="alert alert-info">
+        {% blocktrans trimmed %}
+            Invite another speaker to your submission here. Instead, you can also send them
+            this link:
```
Comment (luto) [inline src/pretalx/cfp/templates/cfp/event/user_submission_invitation.html]: Maybe reword this to make a clear distinction between "**We** send the link here" and "**You** send the link yourself", instead of having "you"/imperative in both cases.

File: src/pretalx/cfp/templates/cfp/event/user_submission_invitation.html (hunk 1 above)
Comment (rixx) [reply src/pretalx/cfp/templates/cfp/event/user_submission_invitation.html]: I normally accidentally click links like that instead. In reality we'd need a "click to copy" link, but I'd open that as an easy issue instead.

File: src/pretalx/cfp/templates/cfp/event/user_submission_edit.html (hunk 2 above)
Comment (rixx) [reply src/pretalx/cfp/templates/cfp/event/user_submission_edit.html]: Nope, you can't, since there is only one shared invitation token per submission. Don't think that's too important either.

File: src/pretalx/submission/models/submission.py (hunk 3 above)
Comment (rixx) [reply src/pretalx/submission/models/submission.py]: Orga invites work entirely differently. Orga invites pre-create a user with the correct permissions. But since we don't have a permission set for speakers as such, I just used a shared invite token.

File: src/pretalx/cfp/templates/cfp/event/user_submission_invitation.html (hunk 4 above)
Comment (rixx) [reply src/pretalx/cfp/templates/cfp/event/user_submission_invitation.html]: Done.

File: src/pretalx/cfp/templates/cfp/event/user_submission_edit.html (hunk 5 above)
Comment (rixx) [reply src/pretalx/cfp/templates/cfp/event/user_submission_edit.html]: Fixed both.

File: src/pretalx/cfp/templates/cfp/event/user_submission_invitation.html (hunk 6 above)
Comment (rixx) [reply src/pretalx/cfp/templates/cfp/event/user_submission_invitation.html]: Tried to clarify

File: src/pretalx/cfp/templates/cfp/event/user_submission_edit.html (hunk 2 above)
Comment (luto) [reply src/pretalx/cfp/templates/cfp/event/user_submission_edit.html]: Oh, I see ;)
